projects/lib/src/board/seirawanboard.cpp

```diff
@@ -0,0 +1,440 @@
+		return WesternBoard::moveFromSanString(str);
+
+	// channeling moves
+	QString s = str.left(index);
+	Move move = WesternBoard::moveFromSanString(s);
+	int promotion = pieceFromSymbol(str.mid(index + 1, 1)).type();
+
+	if (move.isNull() || promotion == 0)
+		return Move();
+
+	Piece piece = Piece(sideToMove(), promotion);
+	if (!reservePieceTypes().contains(piece) ||  reserveCount(piece) < 1)
+		return Move();
+
+	int target = move.targetSquare();
+
+	// castling with channeling onto rook square
+	if (move.sourceSquare() == kingSquare(sideToMove())
+	&&  str.mid(index + 2, 1) == squareString(target).left(1))
+		promotion = rookSquareChanneling(promotion);
+
+	// channeling moves
+	return Move(move.sourceSquare(), target, promotion );
+}
+
+Move SeirawanBoard::moveFromLanString(const QString& str)
+{
+	Move move0(Board::moveFromLanString(str));
+	Side side = sideToMove();
+	int source = move0.sourceSquare();
+	int target = move0.targetSquare();
+	int promotion = move0.promotion();
+
+	// castling move with channeling onto rook square
+	// coded as castling with promotion to auxiliary type
+	if (promotion != Piece::NoPiece
+	&&  pieceAt(source) == Piece(side, Rook)
+	&&  target == kingSquare(side))
+	{
+		return Move(target, source, rookSquareChanneling(promotion));
+	}
+
+	Move move(WesternBoard::moveFromLanString(str));
+	return Move(move.sourceSquare(), move.targetSquare(), promotion);
+}
+
+void SeirawanBoard::vMakeMove(const Move& move, BoardTransition* transition)
+{
+	int source = move.sourceSquare();
+	int prom = move.promotion();
+	Side side = sideToMove();
+
+	WesternBoard::vMakeMove(move, transition);
+
+	if (m_squareMap.contains(source)
+	&&  m_squareMap[source] == 0
+	&&  prom != Piece::NoPiece)
+	{
+		if (prom >= rookSquareChanneling(Hawk))
+		{
+			prom = rookSquareChanneling(prom, backward);
+			source = move.targetSquare();
+		}
+		Piece piece(side, prom);
+		if (reserveCount(piece) > 0)
+		{
+			Square csq = chessSquare(source);
+
+			if (csq.isValid())
+			{
+				removeFromReserve(piece);
+				setSquare(source, piece); // drop-in
+				if (transition != nullptr)
+					transition->addDrop(piece, csq);
+			}
+		}
+	}
+	updateSquareMap(move, +1);
+}
+
+void SeirawanBoard::vUndoMove(const Move& move)
+{
+	Side side = sideToMove();
+	int index = move.sourceSquare();
+
+	if (move.promotion() >= rookSquareChanneling(Hawk))
+		  index = move.targetSquare();
+
+	Piece piece = pieceAt(index);
+
+	updateSquareMap(move, -1);
+
+	// channeling move?
+	if ( piece.side() == side
+	&&   chessSquare(index).isValid()
+	&&   reservePieceTypes().contains(piece))
+	{
+		// undo channeling move
+		addToReserve(piece);
+		// skipped: setSquare(index, Piece());
+	}
+	WesternBoard::vUndoMove(move);
+}
+
+void SeirawanBoard::generateMovesForPiece(QVarLengthArray< Move >& moves,
+					  int pieceType,
+					  int square) const
+{
+	if (!isValidSquare(chessSquare(square)))
+		return;
+
+	QVarLengthArray< Move > moves1, moves2;
+	moves1.clear();
```

Comment:
Aren't `moves1` and `moves2` already empty or cleared after construction?

Reply:
Well spotted.